lib/Doctrine/ODM/MongoDB/Iterator/CachingIterator.php

```diff
@@ -51,8 +50,13 @@ final class CachingIterator implements Countable, Iterator
      */
     public function __construct(Traversable $iterator)
     {
-        $this->iterator = $this->wrapTraversable($iterator);
-        $this->storeCurrentItem();
+        $this->iterator = new IteratorIterator($iterator);
+        $this->iterator->rewind();
+        if ($this->iterator->valid()) {
+            $this->storeCurrentItem();
+        } else {
+            $this->iterator = null;
+        }
```

Comment:
I believe this logic as well as the one contained could be inlined into `storeCurrentItem`. That method already contains logic to deal with an exhausted iterator (through checking the key) and does not store an item in that case - but it also doesn't mark the iterator as exhausted.

Instead of having the separate logic here, it would be nice to inline this logic into `storeCurrentItem` through a guard condition, added as a suggestion below. I've also added suggestions in other places related to this change.

```suggestion
        $this->storeCurrentItem();
```

I can't add comments on `storeCurrentItem` since it wasn't changed, but essentially the guard clause would also set `$this->iterator = null` if the iterator is no longer valid.